adapters/repos/db/vector/hnsw/insert.go

```diff
@@ -105,6 +108,111 @@ func (h *hnsw) AddBatch(ctx context.Context, ids []uint64, vectors [][]float32)
 	return nil
 }
 
+func (h *hnsw) AddMultiBatch(ctx context.Context, docIDs []uint64, vectors [][][]float32) error {
+	if err := ctx.Err(); err != nil {
+		return err
+	}
+	if !h.multivector.Load() {
+		return errors.Errorf("AddMultiBatch called on non-multivector index")
+	}
+	if len(docIDs) != len(vectors) {
+		return errors.Errorf("ids and vectors sizes does not match")
+	}
+	if len(docIDs) == 0 {
+		return errors.Errorf("insertBatch called with empty lists")
+	}
+
+	h.trackDimensionsOnce.Do(func() {
+		atomic.StoreInt32(&h.dims, int32(len(vectors[0][0])))
+	})
+
+	for i, docID := range docIDs {
+		numVectors := len(vectors[i])
+		levels := make([]int, numVectors)
+		for j := range numVectors {
+			levels[j] = int(math.Floor(-math.Log(h.randFunc()) * h.levelNormalizer))
+		}
+
+		h.Lock()
+		counter := h.vecIDcounter
+		h.vecIDcounter += uint64(numVectors)
```

Comment:
What is `vecIDcounter` used for? I don't see it used anywhere else.

Reply:
`vecIDcounter` is used to assign the node id in the graph. The id of the nodes are assigned incrementally and in l.136-l.139 we "block" the ids range [`h.vecIDcounter`, `h.vecIDcounter + numVector`] for a specific document. So that when inserting the vectors belonging to that document we use `nodeID=counter` and then `counter++` for each vector in the same document.

A more clear solution could be to assign to each node `nodeID = h.vecIDcounter` and then `h.vecIDcounter++`, but this would require a lock for each vector we're adding. Instead we "reserve" the ids range one document at time reducing the amount of times we lock the graph from `numDocs * avgVecPerDoc` to `numDocs`.